Find alternative landing spots for blocked reinforcements

abcdefg30 opened the pull request:
Fixes units blocking transports from landing. Easy noticeable on carryalls in D2k (e.g. let the first reinforcements in Atreides04 stay and wait for the second reinforcements).

abcdefg30 committed:
Find alternative landing spots for blocked reinforcements

MustaphaTR commented:
Not tested but i think what this does to move the unit below the carryall. I'm not sure if that's really what we would want to do, it is not the behaviour in vanilla game and only works if blocking actor is mobile, not for structures for example. Carryall should change its drop location if it is blocked.

abcdefg30 commented:
Yes, this does move units below the carryall. I think non-mobile blockers are a separate issue.

pchote commented:
Expanding on @MustaphaTR's suggestion, a `nearEnough` parameter combined with a domain check should be a simple and relatively robust way to have the carryall pick a nearby unblocked cell to land in.

MustaphaTR commented:
You could block carryall with units in original tho, making carryall take back off(or not land at all if it was already blocked) and go to another cell. I would prefer that behaviour to apply OpenRA too. You could technically do that infinitely and make carryall fail to place its units.

abcdefg30 commented:
Updated.

pchote commented on OpenRA.Mods.Common/Scripting/Global/ReinforcementsGlobal.cs:
```diff
@@ -146,14 +148,32 @@ public Actor[] Reinforce(Player owner, string[] actorTypes, CPos[] entryPath, in
+					{
+						foreach (var c in transport.World.Map.FindTilesInCircle(destination, dropRange))
+						{
+							if (!aircraftInfo.CanLand(c))
```
Please also do a `DomainIndex` check to make sure that this isn't on the wrong side of a river/cliff/other obstacle.

abcdefg30 replied:
Added.

abcdefg30 committed:
Add an domain index check

MustaphaTR commented:
Needs rebease.

abcdefg30 commented:
Rebased.

reaperrr commented:
> OpenRA.Mods.Common/Scripting/Global/ReinforcementsGlobal.cs(160,24): error CS1061: Type `OpenRA.Mods.Common.Traits.AircraftInfo' does not contain a definition for `CanLand' and no extension method `CanLand' of type `OpenRA.Mods.Common.Traits.AircraftInfo' could be found. Are you missing an assembly reference?
OpenRA.Mods.Common/Scripting/Global/ReinforcementsGlobal.cs(164,26): error CS1061: Type `OpenRA.Mods.Common.Traits.AircraftInfo' does not contain a definition for `CanLand' and no extension method `CanLand' of type `OpenRA.Mods.Common.Traits.AircraftInfo' could be found.

Rebase error?

abcdefg30 commented:
> Rebase error?

Yes, fixed. Thanks.

RoosterDragon reviewed (dismissed):
:+1:

(Maybe squash the commits? Up to you)

pchote commented on OpenRA.Mods.Common/Scripting/Global/ReinforcementsGlobal.cs:
```diff
@@ -143,17 +151,50 @@ public Actor[] Reinforce(Player owner, string[] actorTypes, CPos[] entryPath, in
+
+							if (cargo != null)
+							{
+								var mobiles = cargo.Passengers.Select(a =>
```
Please pull this outside the `FindTilesInCircle` so that it isn't unnecessarily repeated.

pchote replied:
Something like the following should work:
```diff
--- a/OpenRA.Mods.Common/Scripting/Global/ReinforcementsGlobal.cs
+++ b/OpenRA.Mods.Common/Scripting/Global/ReinforcementsGlobal.cs
@@ -159,25 +159,22 @@ public Actor[] Reinforce(Player owner, string[] actorTypes, CPos[] entryPath, in
                                        // Try to find an alternative landing spot if we can't land at the current destination
                                        if (!aircraft.CanLand(destination) && dropRange > 0)
                                        {
+                                               var mobiles = cargo != null ? cargo.Passengers.Select(a =>
+                                               {
+                                                       var mobileInfo = a.Info.TraitInfoOrDefault<MobileInfo>();
+                                                       if (mobileInfo == null)
+                                                               return new Pair<MobileInfo, uint>(null, 0);
+
+                                                       return new Pair<MobileInfo, uint>(mobileInfo, (uint)mobileInfo.GetMovementClass(a.World.Map.Rules.TileSet));
+                                               }) : new Pair<MobileInfo, uint>[0];
+
                                                foreach (var c in transport.World.Map.FindTilesInCircle(destination, dropRange))
                                                {
                                                        if (!aircraft.CanLand(c))
                                                                continue;
 
-                                                       if (cargo != null)
-                                                       {
-                                                               var mobiles = cargo.Passengers.Select(a =>
-                                                               {
-                                                                       var mobileInfo = a.Info.TraitInfoOrDefault<MobileInfo>();
-                                                                       if (mobileInfo == null)
-                                                                               return new Pair<MobileInfo, uint>(null, 0);
-
-                                                                       return new Pair<MobileInfo, uint>(mobileInfo, (uint)mobileInfo.GetMovementClass(a.World.Map.Rules.TileSet));
-                                                               });
-
-                                                               if (!mobiles.All(m => m.First == null || domainIndex.IsPassable(destination, c, m.First, m.Second)))
-                                                                       continue;
-                                                       }
+                                                       if (!mobiles.All(m => m.First == null || domainIndex.IsPassable(destination, c, m.First, m.Second)))
+                                                               continue;
 
                                                        destination = c;
                                                        break;
```

pchote reviewed (dismissed):
LGTM.  Verified fix with both Atreides04 and a modified version of nod04b.  Just one perf request and then feel free to merge directly.

abcdefg30 commented:
Updated and rebased.

> Just one perf request and then feel free to merge directly.

Can't though as updating dismissed all reviews and you need one review to merge...

reaperrr reviewed (approved):
Code looks good as far as I can tell and reinforcements on RA shellmap show no regressions.

reaperrr commented:
[changelog](https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/583f6a53a787da60ee0edc4021cf3bff244a7783...2c2e15263989b1bc0d92e8f4cf265b47184f8927)
